### ISSUE TITLE
Replace `PyIteratorValue` with distinct iterators for each iterable type

### ISSUE THREAD

**Issue by OddCoincidence** [RustPython/RustPython#692]
We currently have one type called `iter` being used for each iterator. This is incorrect - each type should have its own:

```
>>> iter([])
<list_iterator object at 0x7fe685314cf8>
>>> iter({})
<dict_keyiterator object at 0x7fe685435368>
```

Types to fix:

- [ ] list
- [ ] set
- [ ] dict
- [ ] range
- [ ] bytes
- [ ] tuple

**Comment by skinny121** [comment]
To me, this seems like an implementation-defined detail, although having a different iterator for some collections would be more memory/speed efficient, as the iteration can be optimised for that collection.

Also, you missed tuple.

**Comment by adrian17** [comment]
Can be closed with #772 ?.
(well mostly, as `set` and `dict` use list_iterator for now. But that sounds like a follow-up issue to be done after the dict/set refactor are done)
